File: api/generate.py
"""
Vercel Serverless Function: Video Generation Proxy
POST /api/generate
"""
from http.server import BaseHTTPRequestHandler
import json
import os
import requests


API_ENDPOINT = 'https://api.dzz.ai/v1/chat/completions'


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            # Get API key from environment
            api_key = os.getenv('SORA_API_KEY')
            if not api_key:
                self.send_error(500, 'SORA_API_KEY not configured')
                return
            
            # Parse request body
            content_length = int(self.headers['Content-Length'])
            body = self.rfile.read(content_length)
            data = json.loads(body.decode('utf-8'))
            
            prompt = data.get('prompt', '')
            resolution = data.get('resolution', '1080p')
            aspect_ratio = data.get('aspectRatio', '16:9')
            mode = data.get('mode', 'text')
            image_base64 = data.get('image', None)
            
            # Validation
            if not prompt and mode == 'text':
                self.send_error(400, 'Prompt is required')
                return
            
            if mode == 'image' and not image_base64:
                self.send_error(400, 'Image is required for image-to-video mode')
                return
            
            # Build message content
            if mode == 'image' and image_base64:
                content = [
                    {
                        "type": "image_url",
                        "image_url": {
                            "url": image_base64
                        }
                    }
                ]
                if prompt:
                    content.insert(0, {
                        "type": "text",
                        "text": prompt
                    })
            else:
                content = prompt
            
            # Build API payload
            payload = {
                "messages": [
                    {
                        "role": "user",
                        "content": content
                    }
                ],
                "model": "sora_video2",
                "stream": True
            }
            
            print(f"Generating - resolution: {resolution}, aspect: {aspect_ratio}, mode: {mode}")
            
            # Call upstream API with streaming
            headers = {
                'Authorization': f'Bearer {api_key}',
                'Content-Type': 'application/json'
            }
            
            response = requests.post(
                API_ENDPOINT,
                headers=headers,
                json=payload,
                stream=True,
                timeout=300
            )
            
            if not response.ok:
                self.send_error(response.status_code, f'API error: {response.text[:200]}')
                return
            
            # Stream response back to client
            self.send_response(200)
            self.send_header('Content-Type', 'text/event-stream')
            self.send_header('Cache-Control', 'no-cache')
            self.send_header('Connection', 'keep-alive')
            self.send_header('X-Accel-Buffering', 'no')
            self.end_headers()
            
            # Stream lines from upstream
            for line in response.iter_lines():
                if line:
                    decoded_line = line.decode('utf-8')
                    if decoded_line.startswith('data: '):
                        data_content = decoded_line[6:]
                        self.wfile.write(f"data: {data_content}\n\n".encode('utf-8'))
                        self.wfile.flush()
            
        except Exception as e:
            print(f'Error: {str(e)}')
            self.send_error(500, str(e))

File: functions/api/generate.js
/**
 * EdgeOne Pages Node Function: Video Generation Proxy
 * POST /api/generate
 * 
 * Environment Variables Required:
 * - SORA_API_KEY: Your API key for api.dzz.ai
 */

const API_ENDPOINT = 'https://api.dzz.ai/v1/chat/completions';

export default async function handler(request, context) {
  // Only accept POST
  if (request.method !== 'POST') {
    return new Response(
      JSON.stringify({ error: 'Method not allowed' }),
      { status: 405, headers: { 'Content-Type': 'application/json' } }
    );
  }

  try {
    // Get API key from environment
    const apiKey = context.env.SORA_API_KEY;
    if (!apiKey) {
      return new Response(
        JSON.stringify({ error: 'SORA_API_KEY not configured' }),
        { status: 500, headers: { 'Content-Type': 'application/json' } }
      );
    }

    // Parse request body
    const data = await request.json();
    const { prompt, resolution, aspectRatio, mode, image } = data;

    if (!prompt && mode === 'text') {
      return new Response(
        JSON.stringify({ error: 'Prompt is required' }),
        { status: 400, headers: { 'Content-Type': 'application/json' } }
      );
    }

    if (mode === 'image' && !image) {
      return new Response(
        JSON.stringify({ error: 'Image is required for image-to-video mode' }),
        { status: 400, headers: { 'Content-Type': 'application/json' } }
      );
    }

    // Build message content
    let content;
    if (mode === 'image' && image) {
      // Image-to-video mode
      content = [
        {
          type: 'image_url',
          image_url: {
            url: image
          }
        }
      ];
      if (prompt) {
        content.unshift({
          type: 'text',
          text: prompt
        });
      }
    } else {
      // Text-to-video mode
      content = prompt;
    }

    // Build API payload
    const payload = {
      messages: [
        {
          role: 'user',
          content: content
        }
      ],
      model: 'sora_video2',
      stream: true
    };

    console.log(`Generating video - resolution: ${resolution}, aspect: ${aspectRatio}, mode: ${mode}`);

    // Call upstream API
    const response = await fetch(API_ENDPOINT, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`API error: ${response.status} - ${errorText}`);
      return new Response(
        JSON.stringify({ 
          error: `API error: ${response.status}`,
          details: errorText 
        }),
        { 
          status: response.status,
          headers: { 'Content-Type': 'application/json' }
        }
      );
    }

    // Stream the response back to client
    return new Response(response.body, {
      status: 200,
      headers: {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive',
        'X-Accel-Buffering': 'no'
      }
    });

  } catch (error) {
    console.error('Function error:', error);
    return new Response(
      JSON.stringify({ 
        error: error.message || 'Internal server error',
        stack: error.stack
      }),
      { 
        status: 500,
        headers: { 'Content-Type': 'application/json' }
      }
    );
  }
}

